api/test/e2e/route_service_upstream_test.go

```diff
@@ -0,0 +1,304 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package e2e
+
+import (
+	"net/http"
+	"testing"
+)
+
+func TestRoute_Invalid_Service_And_Service(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "use service that not exist - dashboard",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+				"uri": "/hello_",
+				"service_id": "not-exists"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:   "hit invalid route on data plane",
+			Object:     APISIXExpect(t),
+			Method:     http.MethodGet,
+			Path:       "/hello_",
+			ExpectCode: http.StatusNotFound,
+		},
+		{
+			caseDesc: "use upstream that not exist",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+				"uri": "/hello_",
+				"upstream_id": "not-exists"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:   "hit invalid route on data plane",
+			Object:     APISIXExpect(t),
+			Method:     http.MethodGet,
+			Path:       "/hello_",
+			ExpectCode: http.StatusNotFound,
+		},
+		{
+			caseDesc: "create service and upstream together at the same time",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/routes/r1",
+			Body: `{
+				"uri": "/hello_",
+				"service_id": "not-exists-service",
+				"upstream_id": "not-exists-upstream"
+			}`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+		},
+		{
+			caseDesc:   "hit invalid route on data plane",
+			Object:     APISIXExpect(t),
+			Method:     http.MethodGet,
+			Path:       "/hello_",
+			ExpectCode: http.StatusNotFound,
+		},
+	}
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+}
+
+func TestRoute_Create_Service(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc: "create service",
+			Object:   MangerApiExpect(t),
+			Method:   http.MethodPut,
+			Path:     "/apisix/admin/services/200",
+			Body: `{
+					"upstream": {
+					"type": "roundrobin",
+					"nodes": [
+						{
+							"host": "172.16.238.20",
```

Comment:
the IP `172.16.238.20` is so obscure, it is defined in docker-compose.yaml.
If we do not test in docker container, it is recommended to use port mapping and use IP `127.0.0.1` to prevent network segment conflicts.

Reply:
Thanks for your suggestion, I learned something new, the current e2e uses the docker container provided by github action.